File: javascript/duplicate_encoder.js
/*

input:
  - string

output:
  - encoded string of ")" and "("

rules:
  - each character that appears once in original string is replaced by "("
  - characters that appear more than once are replaced by ")"
  - ignore capitalization when determining if a character is a duplicate

algorithm:
  - create a count object with counts of each character
    - reduce, set to 1 if not present, otherwise add 1
  - map the array of characters, replacing with ")" or "("

*/

let duplicateEncode = function (word) {
  let chars = word.toLowerCase().split('');
  let counts = chars.reduce((counts, character) => {
    counts[character] ? counts[character] += 1 : counts[character] = 1;
    return counts;
  }, {});

  return chars.map(character => {
    if (counts[character] > 1) {
      return ')';
    } else {
      return '(';
    }
  }).join('');
}

console.log(duplicateEncode("din"));
console.log(duplicateEncode("recede"));
console.log(duplicateEncode("Success"));
console.log(duplicateEncode("(( @"));

File: go/duplicate_encoder.go
package main

import (
	"fmt"
	"strings"
)

func DuplicateEncode(word string) (result string) {
	lower := strings.ToLower(word)
	m := make(map[string]int)
	for _, character := range lower {
		m[string(character)]++
	}
	for _, character := range lower {
		if m[string(character)] > 1 {
			result += ")"
		} else {
			result += "("
		}
	}
	return
}

func main() {
	fmt.Println(DuplicateEncode("din"))
	fmt.Println(DuplicateEncode("recede"))
	fmt.Println(DuplicateEncode("(( @"))
	fmt.Println(DuplicateEncode("Success"))
}
